course5/homework10/linked_list.h

```diff
@@ -0,0 +1,221 @@
+    }
+
+    bool operator==(LinkedList& cmp) const
+    {
+        std::cout << mSize << " " << cmp.mSize << std::endl;
+        if (mSize != cmp.mSize)
+            return false;
+        if (data != cmp.data)
+            return false;
+        if (!(next && cmp.next))
+             return true;
+        if  (next != cmp.next)
+            return false;
+        return true;
+    }
+
+    int size() const { return mSize; }
+    T front() const // TODO should this be a pointer? const?
+    {
+        if (data)
+            return *data.get();
+        throw std::range_error("Calling front on empty LinkedList is undefined.");
+    }
+    void push_front(const T& val)
+    {
+        if (mSize > 0)
+        {
+            LinkedList<T> new_node;
+            new_node.data = std::move(this->data);
+            new_node.next = std::move(this->next);
+            data = std::make_unique<T>(val);
+            next = std::make_unique<LinkedList<T>>(std::move(new_node));
+        }
+        else
+        {
+            data = std::make_unique<T>(val);
+        }
+        mSize++;
+    }
+
+    template< class... Args >
+    void emplace_front(Args&&... args)
+    {
+        if (mSize > 0)
+        {
+            LinkedList<T> new_node;
+            new_node.data = std::move(this->data);
+            new_node.next = std::move(this->next);
+            data = std::make_unique<T>(args...);
+            next = std::make_unique<LinkedList<T>>(std::move(new_node));
+        }
+        else
+        {
+            data = std::make_unique<T>(args...);
+        }
+        mSize++;
+    }
+
+    void pop_front()
+    {
+        if (next)
+        {
+            LinkedList<T> new_head = std::move(*next.get());
+            data = std::move(new_head.data);
+            next = std::move(new_head.next);
+            mSize--;
+        }
+    }
+
+    void print() const
+    {
+        std::cout << "------------------------------------------------------------------------------" << std::endl;
+        std::cout << "getting this" << std::endl;
+        LinkedList<T> current = *this;
+        std::cout << "got this" << std::endl;
+        for (auto i = 0; i < mSize; i++)
+        {
+            std::cout << i << " [" << current.front() << "]" << std::endl;
+            if (current.next)
+            {
+                std::cout << "getting next ptr" << std::endl;
+                LinkedList<T> *newcurrent = current.next.get();
+                std::cout << "getting next current" << std::endl;
+                current = *(newcurrent);
+                std::cout << "got next" << std::endl;
+            }
+            else
+            {
+                std::cout << "breaking" << std::endl;
+                break;
+            }
+        }
+        std::cout << std::endl;
+    }
+
+    iterator begin() { return iterator(*this); }
+    iterator end() { return iterator(nullptr); }
+
+private:
+    std::unique_ptr<T> data;
+    std::unique_ptr<LinkedList<T>> next;
```

Comment:
I think the entire architecture of the list is wrong. I suggest you take a pencil, paper, and draw it.